File: rust/src/benchmarks/matmul8t.rs
use super::super::{Benchmark, INPUT, helper};
use rayon::ThreadPoolBuilder;
use rayon::prelude::*;

pub struct Matmul8T {
    n: i32,
    result: u32,
}

impl Matmul8T {
    pub fn new() -> Self {
        let name = "Matmul8T".to_string();
        let iterations: i32 = INPUT.get()
            .unwrap()
            .get(&name)
            .and_then(|s| s.parse().ok())
            .unwrap_or(0);
        
        Self {
            n: iterations,
            result: 0,
        }
    }

    fn matgen(&self, n: usize) -> Vec<Vec<f64>> {
        let tmp = 1.0 / (n * n) as f64;
        let mut a = vec![vec![0.0; n]; n];
        
        for i in 0..n {
            for j in 0..n {
                a[i][j] = tmp * (i as f64 - j as f64) * (i as f64 + j as f64);
            }
        }
        
        a
    }

    fn matmul_parallel(&self, a: &[Vec<f64>], b: &[Vec<f64>]) -> Vec<Vec<f64>> {
        let pool = ThreadPoolBuilder::new()
            .num_threads(8)
            .build()
            .expect("Failed to create local thread pool 8");
        
        let size = a.len();
        
        // Транспонируем b
        let mut b_t = vec![vec![0.0; size]; size];
        for i in 0..size {
            for j in 0..size {
                b_t[j][i] = b[i][j];
            }
        }
        
        // Умножение матриц
        let mut c = vec![vec![0.0; size]; size];
                
        pool.install(|| {
            c.par_iter_mut().enumerate().for_each(|(i, ci)| {
                let ai = &a[i];
                for j in 0..size {
                    let mut sum = 0.0;
                    let b_tj = &b_t[j];
                    
                    for k in 0..size {
                        sum += ai[k] * b_tj[k];
                    }
                    
                    ci[j] = sum;
                }
            });
        });
        
        c
    }
}

impl Benchmark for Matmul8T {
    fn name(&self) -> String {
        "Matmul8T".to_string()
    }
    
    fn iterations(&self) -> i32 {
        self.n
    }
    
    fn run(&mut self) {
        let n = self.n as usize;
        
        let a = self.matgen(n);
        let b = self.matgen(n);
        
        let c = self.matmul_parallel(&a, &b);
        
        let center_value = c[n >> 1][n >> 1];
        self.result = helper::checksum_f64(center_value);
    }
    
    fn result(&self) -> i64 {
        self.result as i64
    }
}

File: csharp/Matmul8T.cs
using System.Threading.Tasks;

public class Matmul8T : Benchmark
{
    private int _n;
    private uint _result;

    public Matmul8T()
    {
        _result = 0;
        _n = (int)ConfigVal("n");
    }

    public override void Run(long IterationId)
    {
        double[][] a = MatGen(_n);
        double[][] b = MatGen(_n);
        double[][] c = MatMulParallel(a, b);

        double value = c[_n >> 1][_n >> 1];
        _result += Helper.Checksum(value);
    }

    public override uint Checksum => _result;

    private double[][] MatMulParallel(double[][] a, double[][] b)
    {
        int size = a.Length;

        double[][] bT = new double[size][];
        for (int i = 0; i < size; i++)
        {
            bT[i] = new double[size];
            for (int j = 0; j < size; j++)
            {
                bT[i][j] = b[j][i];
            }
        }

        double[][] c = new double[size][];

        for (int i = 0; i < size; i++)
        {
            c[i] = new double[size];
        }

        Parallel.For(0, size, new ParallelOptions { MaxDegreeOfParallelism = 8 }, i =>
        {
            double[] ai = a[i];
            double[] ci = c[i];

            for (int j = 0; j < size; j++)
            {
                double sum = 0.0;
                double[] bTj = bT[j];

                for (int k = 0; k < size; k++)
                {
                    sum += ai[k] * bTj[k];
                }

                ci[j] = sum;
            }
        });

        return c;
    }

    private double[][] MatGen(int n)
    {
        double tmp = 1.0 / n / n;
        double[][] a = new double[n][];

        for (int i = 0; i < n; i++)
        {
            a[i] = new double[n];
            for (int j = 0; j < n; j++)
            {
                a[i][j] = tmp * (i - j) * (i + j);
            }
        }

        return a;
    }
}